### PR TITLE
Support configuring ip_source in Caddyfile.

### PR DESCRIPTION
**Pull request by caiych**

### CODE REVIEW

File: caddyfile.go [hunk 1]
```diff
@@ -20,6 +20,8 @@ func init() {
 //         }
 //         check_interval <duration>
 //         provider <name> ...
+//         ip_source upnp
+//         ip_source endpoint https://icanhazip.com
```
Comment (francislavoie) [inline caddyfile.go:24]: I think I would write this as `ip_source upnp|endpoint <endpoint>`

File: README.md [hunk 2]
```diff
@@ -85,6 +85,8 @@ Equivalent Caddyfile (there is not currently a way to customize IP sources via C
 			example.net subdomain
 		}
 		check_interval 5m
+		ip_source upnp
+		ip_source endpoint https://icanhazip.com
```
Comment (francislavoie) [inline README.md:89]: I think it would be helpful to show that multiple IP source endpoints can be configured. Can you add another here?

**Comment by fu-sen** [comment]
Since `https://api.ipify.org` is from AWS in the US, the response is poor depending on the region.
This response was ideal because `https://icanhazip.com` used Cloudflare.

File: caddyfile_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+func equivalentJSON(s1, s2 string, t *testing.T) string {
+	var v1, v2 map[string]interface{}
+	if err := json.Unmarshal([]byte(s1), &v1); err != nil {
+		t.Error(err)
+	}
+	if err := json.Unmarshal([]byte(s2), &v2); err != nil {
+		t.Error(err)
+	}
+
+	return cmp.Diff(v1, v2)
```
Comment (mholt) [inline caddyfile_test.go:98]: Can we drop this dependency, maybe `reflect.DeepEqual` or something instead?

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for the PR, and the test cases.

Since the IP source is modular, its Caddyfile parsing would best be modular, too. You can see several examples in the Caddy code base how this is done, for example, the reverse_proxy's transport modules: https://github.com/caddyserver/caddy/blob/124ba1ba714220322c65625012c564f89b75bfc9/modules/caddyhttp/reverseproxy/caddyfile.go#L631-L648

```go
			case "transport":
				if !d.NextArg() {
					return d.ArgErr()
				}
				if h.TransportRaw != nil {
					return d.Err("transport already specified")
				}
				transportModuleName = d.Val()
				modID := "http.reverse_proxy.transport." + transportModuleName
				unm, err := caddyfile.UnmarshalModule(d, modID)
				if err != nil {
					return err
				}
				rt, ok := unm.(http.RoundTripper)
				if !ok {
					return d.Errf("module %s (%T) is not a RoundTripper", modID, unm)
				}
				transport = rt
```

File: caddyfile_test.go (hunk 3 above)
Comment (caiych) [reply caddyfile_test.go:103]: We can, but I would like to suggest adopting this package. I have had good experience with this package in other projects and I personally recommend using it.

It's intended to be a better alternative of reflect.DeepEqual and support versatile transformations providing readable error messages in tests: https://pkg.go.dev/github.com/google/go-cmp/cmp

If you insist removing this, let me know and I'll remove it.

Thanks.

**Comment by caiych** [comment]
Updated to use a modular parsing. Also changing the syntax of Caddyfile by requiring a "simple_http" arg before endpoint.

I'm not sure how to do this to merge endpoint from multi-line, so after this, multiple "ip_source simple_http endpoint" will result in multiple SimpleHTTP objects -- not sure if that is a concern(performance?)

File: ipsource.go [hunk 4]
```diff
@@ -57,6 +61,23 @@ func (SimpleHTTP) CaddyModule() caddy.ModuleInfo {
...
+		unused   string
+		arg      string
+		endpoint string
+	)
+	if !d.AllArgs(&unused, &arg, &endpoint) {
+		return d.ArgErr()
+	}
+	if arg != "endpoint" {
+		return fmt.Errorf("wrong arg receieved: %v, only support \"endpoint\"", arg)
+	}
```
Comment (francislavoie) [inline ipsource.go:75]: I think `endpoint` can be dropped now, no use to have that as the first arg.

File: caddyfile_test.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+				ip_source simple_http endpoint http://1.com
+				ip_source simple_http endpoint http://2.com
+			}`),
+			want: ` {
+				"ip_sources": [
+					{
+						"source": "simple_http",
+						"endpoints": ["http://1.com"]
+					},
+                    {
```
Comment (francislavoie) [inline caddyfile_test.go:46]: This line has spaces instead of tabs
```suggestion
					{
```

**Review by mholt — APPROVED** [review]
Great, that looks much better. Thanks for the contribution!

**Pull request merged**